[WIP] LDA tutorial, tips and tricks

olavurmortensen opened the pull request:
@tmylk @piskvorky A tutorial on LDA sharing some of my experience, as requested.

@tmylk I'm sure you have some comments on it. Thought it would be easiest with a PR. It's still a work in progress, as reflected by the "TODO" list in the start of the tutorial.

Not exactly sure what you would like the tutorial to be, but I tried to explain what the goal of it was in the introduction.

tmylk commented:
Could you link to it in `tutorials.md` file as well?

olavurmortensen commented:
Will do @tmylk. Does that mean you think its ok as it is? In that case I'll just clean it up one of these days to prepare it for merging.

tmylk commented:
@olavurmortensen When do you think this would be finished?

olavurmortensen commented:
@tmylk Well, I thought you would have some comments. If you do not, then I think I can finish it tomorrow.

piskvorky commented on docs/notebooks/LDA_tutorial.ipynb:
```diff
@@ -0,0 +1,732 @@
+    "\n",
+    "> **Note:**\n",
+    ">\n",
+    "> This tutorial uses the scikit-learn and nltk libraries, although you can replace them with others if you want. Python 3 is used, although Python 2.7 can be used as well.\n",
```
Where is sklearn used? I cannot find it.

piskvorky commented on docs/notebooks/LDA_tutorial.ipynb:
```diff
@@ -0,0 +1,732 @@
+    "* Check out a RaRe [blog post](http://rare-technologies.com/what-is-topic-coherence/) on the AKSW topic coherence measure.\n",
+    "* [pyLDAvis](https://pyldavis.readthedocs.io/en/latest/index.html).\n",
+    "* Read some more [Gensim tutorials](https://github.com/RaRe-Technologies/gensim/blob/develop/tutorials.md#tutorials).\n",
+    "* If you haven't already, read [1] and [2] (see references)."
```
Normal hyperlinks better. Or at least hyperlink the references.

cscorley commented:
Briefly looked through it this morning, all I can add right now would be an explanation as to why 5 models trained with exactly the same input would have different output, e.g., perplexity, and perhaps an aside on how to achieve 1:1 training with the random state parameter.

olavurmortensen commented:
@cscorley The 5 models are different because of random initialization (specifically, random initialization of some hyperparameters, e.g. gamma). But you bring up an important point that this should be explained in the tutorial, and maybe even set the random state just make it more explicit.

olavurmortensen commented:
I have updated the tutorial according to the comments, thanks @piskvorky and @cscorley. Also added a link in `tutorials.md`.

I also changed the name because, bizarrely, someone posted a tutorial with the same name I was using just a week ago.

@tmylk and @piskvorky Will the tutorial appear on the RaRe blog?

olavurmortensen commented:
@tmylk @piskvorky Ready for merging. The conflict is because I changed the `tutorials.md` file, I think.

When it's merged I'll submit a blog post on WordPress for review as well.

tmylk commented on docs/notebooks/lda_training_tips.ipynb:
```diff
@@ -0,0 +1,726 @@
+   "source": [
+    "# LDA: training tips\n",
+    "\n",
+    "LDA is a probabilistic hierarchical Bayesian model that is a mixture model as well as a mixed membership model... but we won't be getting into any of that.\n",
```
The first sentence should say what this tutorial is. Here it is done in the Nth sentence - please move it to the very first line. It is ok to discuss what this tutorial is not, but it should be later

olavurmortensen replied:
True. The first sentence is also a tad snarky. I just removed the first sentence, does anything else need to be changed in that regard?

tmylk replied:
"In this tutorial I will show how to pre-process text and train LDA on it"

olavurmortensen replied:
Better now?

tmylk commented on docs/notebooks/lda_training_tips.ipynb:
```diff
@@ -0,0 +1,726 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We select the model with the lowest perplexity."
```
why do you do that? you talk about topic coherence later so it is confusing

olavurmortensen replied:
Yes, I agree this is not a good way of selecting a model at all.

olavurmortensen replied:
I removed the sections about model selection.

tmylk replied:
Thanks

tmylk commented on docs/notebooks/lda_training_tips.ipynb:
```diff
@@ -0,0 +1,726 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "[pyLDAvis](https://pyldavis.readthedocs.io/en/latest/index.html) can be fun and useful. Include the code below in your notebook to visualize your topics with pyLDAvis.\n",
```
Please add actual pyLDAVis output to the notebook

olavurmortensen replied:
I removed the text about it. Come to think of it, since it's mentioned in other RaRe blogs, there isn't much need for it in this one.
